### PR TITLE
feat(finance,reporting): report contracts in progress and unperformed services

### PR DESCRIPTION
**Pull request by mihaipxm**
Romanian tour operators file a periodic return on contracts in progress: how many, what they are worth in lei, and how much has been collected against them. Every structural input was present and there was no report, so assembling one month meant reading figures off PDFs for a day.

**No new page.** It arrives as **one dataset, five widgets and one template** on the existing reports surface, which an operator adds to their own collection through the template route that already ships.

## Why a dataset was the missing piece

More of this was already built than the issue assumed. Custom widgets (`reportWidgetInstance` has a `kind: "custom"` variant), saved reports, `POST /templates/:id/instantiate`, and csv/xlsx/pdf export all shipped.

What was missing was **data those widgets could ask for**. There was exactly one dataset in the repo — `finance.receivables`, invoice grain, dated on `issueDate`. Nothing exposed `booking_items.service_date`, `bookings.confirmed_at`, booking value, or collections at *payment* date. And the query language is deliberately *"SQL-like … but has no joins, subqueries, statements, or table access"*, so an operator could not assemble them however good the builder got — and the totals and the line list have to come from the **same** dataset.

## The dataset

`finance.unperformed-services` — booking grain, spanning three (booking value, booking-item service dates, payment collections) with collections pre-aggregated and refunds netted inside it. Finance owns it because it needs both sides and `finance→bookings` is an established read pair with 29 baselined reach-ins, while `bookings→finance` is not a pair at all and `verify:table-privacy` forbids opening one.

Three decisions worth review:

**Both readings of "advance" ship, labelled.** The issue is explicit that reading it strictly (money against a balance still owed) versus as all collections on a running contract changes the total materially. Rather than pick, the dataset exposes `advancesStrictReportingCents` and `collectionsTotalReportingCents` side by side so an operator filing a return chooses knowingly instead of discovering later which one the platform picked.

**Nothing is converted at read time.** Values come from the rate each contract's own documents were stamped with (#4712). A contract nothing has stamped reports `null` and the result carries a warning naming how many rows are excluded — because looking a rate up now is exactly what the FX work exists to prevent, and a total that silently omits rows reads as complete when it is short.

**The count is booking-derived, and says so.** `contracts` holds 225 rows of which 14 are `issued`, so a contract-record count would be quietly wrong. Every label says booking.

## Three changes to the reporting feature

Without these a parameterised template exists but an operator cannot drive it.

**Template parameters are descriptors, not names.** `parameters` was `string[]` — no type, label, default or required flag, so nothing could render a period picker. Now `{ id, label, description?, valueType, required, defaultValue? }`, carried through the deployment graph (whose validator checks the shape), and instantiating a template seeds its declared defaults so the report opens showing something rather than a wall of missing-parameter errors.

**Export accepts parameters from the caller.** The route passed `{}`, so exporting a different period meant editing and saving the report first — leaving it permanently describing whichever period was exported last. The export is the artifact that gets filed, so this mattered.

**A dataset may own its period.** `applyPageDateWindow` injects one field with both bounds:

```
field >= dateFrom AND field <= dateTo
```

This period is two fields with *opposite open bounds* — concluded on or before period end (no lower bound), services running on or after period start (no upper bound). Not expressible, and wiring it up would answer a different question while looking like it answered this one. Datasets that do not fit declare no `defaultDateField`, read their own parameters, and **fail when absent** rather than silently covering all time. Written up in `docs/architecture/reporting-datasets.md`.

## One refactor

Finance's report query compiler is now shared between both datasets rather than copied. The alternative was not one dataset with a bespoke compiler; it was every dataset with its own copy of the same grouping, aliasing and filter rules, diverging one fix at a time. A dataset owns its relation, its fields and its own answerability rule — receivables keeps its "group by currency or filter to one" invariant as a hook it passes in. Its nine existing unit tests are the regression net and pass unchanged.

## The open question I did not guess at

"In progress" has two readings — still running **at** period end, versus running at any point **during** it — and they give different numbers. The issue's own cancelled-departure case only makes sense under the second, so that is what is implemented and what the tests pin. **The offered worked month should settle it**; whichever wins gets written into the dataset's `grain` string, where a report reader can see it.

## Verification

- **The integration test drives the real path over real Postgres** and is registered by name in the `integration` CI lane — an unlisted integration test here is green by never running, which for a regulatory report would be the worst place for that to happen. It covers each edge case the issue asked to have encoded: a cancelled, refunded departure inside the period contributing contracts and value but zero advances; collections counted only up to period end; a contract collected in full distinguishing the two advance readings; an unstamped contract reporting null with a warning; and a fiscal invoice with no payment recorded against it being flagged rather than assumed collected.
- Unit coverage for the compiler (period binding, refusal without a period, money columns denominated by the dataset's own currency dimension) and for both reporting-feature changes.
- The two existing finance reporting tests that assumed a single dataset were generalised rather than renumbered — the rule they encode is still right, they just needed to know each dataset's own currency dimension.
- Full builds and suites for framework (426), finance (646), reporting (22), reporting-contracts (14); `biome check .` clean; **all 94 `verify:architecture` checks pass**.

A first run surfaced a real graph error worth noting: the template originally reused the dataset's id, and graph entity ids are namespaced across datasets, widgets and templates alike. The template is `finance.contracts-in-progress`.

Refs #4704

### CODE REVIEW

File: packages/finance/src/reporting-unperformed-services.ts [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+    ) service ON true
+    LEFT JOIN LATERAL (
+      SELECT
+        -- Collections net of reversals, at each payment's own stamped rate.
+        -- A departure cancelled and refunded inside the period therefore
+        -- contributes its contract and value but nothing to advances.
+        coalesce(sum(
+          CASE payment.status
+            WHEN 'completed' THEN payment.reporting_amount_cents
+            WHEN 'refunded' THEN -payment.reporting_amount_cents
```
[the remark this answers is not in the document]
Comment (mihaipxm) [reply packages/finance/src/reporting-unperformed-services.ts]: Confirmed and fixed in b578af7347.

Checked it rather than took it: the only writers of `payments` in the package are the fx-stamp service, the proforma conversion and `updatePayment` — nothing in the refund path flips a payment to `refunded`. So a settled refund left the collection standing, which is precisely the case the return exists to get right and the one the issue names as an edge case.

Settled `refund_settlements` are now netted, converted at the rate of the payment they reverse (`amount_cents × payment.reporting_amount_cents / payment.amount_cents`) so a full refund cancels exactly and no new stamp is invented — `refund_settlements` carries no reporting column of its own. Payments already marked `refunded` are excluded from the settlement leg so a reversal is never counted twice, and only settlements `settled` on or before period end count.

Your point about the fixture is the one that stung, and it is fixed too: the new cases drive settled, pending and after-period-end settlements against a `completed` payment rather than hand-inserting a synthetic refunded row.

File: packages/finance/src/reporting-unperformed-services.ts [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+          END
+        ), 0)::bigint AS collections_cents,
+        -- A fiscal invoice means the money was taken. Counted separately so a
+        -- gap against recorded payments is visible rather than assumed away.
+        coalesce(sum(
+          CASE
+            WHEN invoice.invoice_type = 'invoice'
+              AND invoice.status <> 'void'
+              AND invoice.base_total_cents IS NOT NULL
+              THEN invoice.base_total_cents
```
[the remark this answers is not in the document]
Comment (mihaipxm) [reply packages/finance/src/reporting-unperformed-services.ts]: Confirmed and fixed in b578af7347.

Right, and worse than the description: with three installments the invoice total was counted three times, so `invoicedNotCollectedReportingCents` reported a gap of twice the contract value on a contract that was fully paid. My fixture only ever had zero or one payment per invoice, so it never saw it.

Collections and invoiced value are now aggregated in separate LATERALs at their own grains.

One thing that fell out of fixing it, worth flagging because it is a judgement call rather than a mechanical fix: converting an invoice whole and converting its installments separately do not agree to the minor unit — 90,000 EUR at 5.352144 is 481,693 as a whole and 481,692 as three parts. Left alone, every installment-paid contract would carry a permanent one-cent "missing payment record" flag, which teaches an operator to ignore the flag that exists to be read. The comparison now allows one minor unit per payment, which is the actual bound on accumulated rounding and needs no assumption about currency decimals. Pinned by a three-installment case asserting the gap is zero.

File: packages/finance/src/reporting-unperformed-services.ts [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+ * complete when it is short. Say so instead — and the remedy is the FX stamp
+ * routes from voyant#4703, not a rate invented here.
+ */
+function unconvertedWarnings(
+  rows: readonly Record<string, unknown>[],
+  columns: ReportResult["columns"],
+): string[] {
+  const converted = columns.find((column) => column.id.includes("Reporting"))
+  if (!converted) return []
+  const unconverted = rows.filter((row) => row[converted.id] === null).length
```
[the remark this answers is not in the document]
Comment (mihaipxm) [reply packages/finance/src/reporting-unperformed-services.ts]: Confirmed and fixed in b578af7347.

This was the one that actually undermined the point of the feature, so thank you for it. The warning exists so a converted total is never quietly short, and it could not fire on the KPI widgets — the only place a total is read. `sum` skips nulls, the compiler coalesces an all-null sum to zero, and the row-level null check was looking at aggregated output that by construction contains no null. My integration coverage exercised the line-list shape only, which is why it passed.

The executor now asks the relation instead of the returned rows: a companion `count(*) FILTER (WHERE "contractValueReportingCents" IS NULL)` over the same parameter-bound relation. It is skipped when the query touches no converted column, because a contract count cannot be shortened by stamping — so a pure count widget still costs one query.

Pinned both ways: an aggregate case asserting the warning fires while every returned row is non-null (which is exactly what the old check would have seen and passed on), and a count case asserting no warning and no second query.

File: packages/reporting/src/service.ts [hunk 4]
```diff
@@ -74,14 +74,22 @@ export function createReportingService(registry: ReportingRegistry) {
     ) {
       const template = registry.getTemplate(input.templateId, input.version)
       if (!template) throw new ReportingRecordNotFoundError("Report template")
+      // Seed the template's declared defaults so an instantiated report opens
+      // showing something. A parameterised template that lands with an empty
+      // bag renders every widget as a missing-parameter error, which reads as
+      // "this template is broken" rather than "choose a period".
+      const parameters: ReportParameters = {}
+      for (const parameter of template.parameters) {
+        if (parameter.defaultValue !== undefined) parameters[parameter.id] = parameter.defaultValue
```
[the remark this answers is not in the document]
Comment (mihaipxm) [reply packages/reporting/src/service.ts]: Confirmed and fixed in b578af7347.

Verified before acting: `ReportBuilderAdmin` reads only `parameters.dateFrom` / `parameters.dateTo` / `parameters.reportCurrency`, and `exportReport` built its URL from `format` alone. So I had shipped the contract, the graph facet, the service and the route, and stopped one layer short of the operator actually being able to use any of it — a template whose every widget errors on a missing period is not a feature.

The builder now reads the declared parameters off the catalog template that the report records in `sourceTemplateId`/`sourceTemplateVersion`, and renders one labelled input per parameter, typed from `valueType` (`date` gets a date input), with the description as a hint and a required marker while empty. Writing goes into `draft.parameters` through the same `updateDraft` path as the existing controls, so it saves and previews like everything else.

`exportReport` now takes the current parameters and puts them on the query string. It was previously relying on `doc.flush()` having landed first for the saved draft to be the exported period; sending them explicitly removes that ordering dependency and is what makes the route change reachable from the product rather than only from an API client.

**Comment by mihaipxm** [comment]
Pushed b578af7347 — all four review findings fixed, plus the CI failure, which was not this PR.

**The CI failure was already red on main.** `tests` failed on two MCP context ratchets in `apps/operator`. I reproduced both on a pristine `origin/main` worktree at `51a87d54dd`, byte-identical: `20,068` and `41,103`. Attribution from the per-tool breakdown:

| tool | was | now | from |
|---|---|---|---|
| `book_product` | 9,700 | 10,999 | #4698 — `externalInvoice` + `suppressDocuments` and their `.describe()` text |
| `record_payment` | 4,538 | 4,912 | #4712 — the three `reporting_*` columns on the payment contract |
| `issue_invoice_from_booking` describe | — | 41,103 | #4698 — `invoiceFromBookingSchema` |

Neither is a new eager tool: still 7 entries, still the same two deliberate domain writes. Both PRs landed while `operator#test` was a turbo cache hit — the same way the #4592 adjustment did, which the test file already records, and which is why an unrelated PR is the one that found it. Ceilings raised to 21,500 and 44,000 with the measurement and attribution written next to them. I did flag rather than normalise the second one: ~10,000 tokens for a single `describe_tool` call, four times the next heaviest, is worth its own piece of work.

**The four findings were all real and all mine.** The two I would call out:

The **aggregate warning** one undermined the point of the feature — the "never quietly short" warning could not fire on the KPI widgets, the only place a total is read, because `sum` skips nulls and the compiler coalesces an all-null sum to zero. It now asks the relation with a companion count.

The **UI** one is the one I should have caught: I shipped the contract, the graph facet, the service and the route, and stopped one layer short of an operator being able to supply a period at all. The builder now renders a template's declared parameters as typed inputs.

Fixing the invoice fan-out also surfaced a judgement call worth a look — converting a whole and converting its parts round differently, so the missing-payment flag now allows one minor unit per payment rather than flagging every installment contract by a cent. Details on that thread.

Re-verified: operator suite 15/16 files (1 skipped), finance 647, reporting 22, reporting-react 53, framework 426, the 14-case integration suite against Postgres, `biome check .` clean, and all 94 `verify:architecture` checks.

**Comment by mihaipxm** [comment]
CI: the ratchet fix worked — `tests`, `architecture`, `build-graph`, `quality-graph`, `operator-image-smoke`, `packaged-acceptance` all green on the previous push. The one remaining failure was `db-integration`, and it was a flake with a removable cause.

**Diagnosis.** It failed on `packages/commerce/tests/integration/promotion-created-command.test.ts` with `Hook timed out in 10000ms` — a file this branch does not touch, and which runs at position 6 in the lane, well before anything this branch adds at position ~24. Re-running the identical commit passed, confirming the flake.

I did not stop there, because the local reproduction was misleading in an instructive way: the test fails locally on my branch *and* would on main, since a bare `TRUNCATE` of the operator schema takes **43 seconds** on Docker-for-Mac. Local said nothing about the branch; only CI did.

**Cause.** These tests share one database and `cleanupTestDb` truncates all 464 tables. Commerce ran them on vitest's defaults — parallel files, 10s hook timeout — so two files TRUNCATE each other and block until a hook gives up, and even an uncontended truncate on a cold database is marginal against 10s.

`packages/finance/vitest.config.ts` and `packages/relationships/vitest.config.ts` already set `fileParallelism: false`, `maxWorkers: 1`, `hookTimeout: 60000` for exactly this. **Commerce was the last package in the lane still on the defaults**, which is why its file is the one that failed. `f822e75ed2` matches the other two rather than adding a per-file exception to a third. Commerce's 187 tests take ~99s serialized.

Worth noting for whoever reviews the ratchet raise in the previous commit: that one was genuinely red on main, and this one was genuinely a flake. Two different answers to "is the red mine?", and both needed checking rather than assuming.
